lib/runtime-c-api/src/metering.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{
+    error::update_last_error, instance::wasmer_instance_t, module::wasmer_module_t, wasmer_result_t,
+};
+use std::slice;
+
+#[cfg(feature = "metering")]
+use wasmer_runtime_core::backend::Compiler;
+
+/// Creates a new Module with gas limit from the given wasm bytes.
+///
+/// Returns `wasmer_result_t::WASMER_OK` upon success.
+///
+/// Returns `wasmer_result_t::WASMER_ERROR` upon failure. Use `wasmer_last_error_length`
+/// and `wasmer_last_error_message` to get an error message.
+#[allow(clippy::cast_ptr_alignment)]
+#[cfg(feature = "metering")]
+#[no_mangle]
+pub unsafe extern "C" fn wasmer_compile_with_limit(
+    module: *mut *mut wasmer_module_t,
+    wasm_bytes: *mut u8,
+    wasm_bytes_len: u32,
+    gas_limit: u64,
+) -> wasmer_result_t {
+    let bytes: &[u8] = slice::from_raw_parts_mut(wasm_bytes, wasm_bytes_len as usize);
+    let result = wasmer_runtime_core::compile_with(bytes, &get_metered_compiler(gas_limit));
+    let new_module = match result {
+        Ok(instance) => instance,
+        Err(error) => {
+            update_last_error(error);
+            return wasmer_result_t::WASMER_ERROR;
+        }
+    };
+    *module = Box::into_raw(Box::new(new_module)) as *mut wasmer_module_t;
+    wasmer_result_t::WASMER_OK
+}
+
+#[cfg(feature = "metering")]
+fn get_metered_compiler(limit: u64) -> impl Compiler {
+    use wasmer_middleware_common::metering;
+    use wasmer_runtime_core::codegen::{MiddlewareChain, StreamingCompiler};
+    use wasmer_singlepass_backend::ModuleCodeGenerator as SinglePassMCG;
+    let c: StreamingCompiler<SinglePassMCG, _, _, _, _> = StreamingCompiler::new(move || {
+        let mut chain = MiddlewareChain::new();
+        chain.push(metering::Metering::new(limit));
+        chain
+    });
+    c
+}
+
+// returns gas used
+#[allow(clippy::cast_ptr_alignment)]
+#[no_mangle]
+#[cfg(feature = "metering")]
+pub unsafe extern "C" fn wasmer_instance_get_points_used(instance: *mut wasmer_instance_t) -> u64 {
+    use wasmer_middleware_common::metering;
+    let instance = &*(instance as *const wasmer_runtime::Instance);
+    let points = metering::get_points_used(instance);
+    points
+}
+
+// sets gas used
+#[allow(clippy::cast_ptr_alignment)]
+#[no_mangle]
+#[cfg(feature = "metering")]
+pub unsafe extern "C" fn wasmer_instance_set_points_used(
+    instance: *mut wasmer_instance_t,
+    new_gas: u64,
+) {
+    use wasmer_middleware_common::metering;
+    let instance = &mut *(instance as *mut wasmer_runtime::Instance);
+    metering::set_points_used(instance, new_gas)
+}
+
+/*** placeholder implementation if metering feature off ***/
+
+// Without metering, wasmer_compile_with_limit is a copy of wasmer_compile
```

Comment:
I'm not sure we need fallbacks. If metering feature is on, then `wasmer_compile_with_limit` will exist, otherwise it will be absent.

Reply:
See above point about the bindings - the ABI is copied into the other repo. I feel this option is more a way to merge for now without it being fully tested, then metering can be always on (remove conditions) when it is ready to release.

If this is fully supported in master of `go-ext-wasmer`, then it should be also in the default binary builds of the .so files

Reply:
To elaborate, cgo will panic on link error (no build) if the linked dll doesn't have the referenced methods. This is why I need stubs (go doesn't have smart feature flags)

Reply:
I don't think that ` wasmer_compile_with_limit` should return without error if the feature is disabled. At least the function should inform the user that it can't enforce a limit.